data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/BinlogEventListener.java

```diff
@@ -343,31 +354,36 @@ private void handleRowChangeEvent(com.github.shyiko.mysql.binlog.event.Event eve
         final long eventTimestampMillis = event.getHeader().getTimestamp();
 
         final BufferAccumulator<Record<Event>> bufferAccumulator = BufferAccumulator.create(buffer, DEFAULT_BUFFER_BATCH_SIZE, BUFFER_TIMEOUT);
-        for (Object[] rowDataArray : rows) {
-            final Map<String, Object> rowDataMap = new HashMap<>();
-            for (int i = 0; i < rowDataArray.length; i++) {
-                final Map<String, String> tbColumnDatatypeMap = dbTableMetadata.getTableColumnDataTypeMap().get(tableMetadata.getFullTableName());
-                final String columnDataType = tbColumnDatatypeMap.get(columnNames.get(i));
-                final Object data =  DataTypeHelper.getDataByColumnType(MySQLDataType.byDataType(columnDataType), columnNames.get(i),
-                        rowDataArray[i], tableMetadata);
-                rowDataMap.put(columnNames.get(i), data);
-            }
 
-            final Event dataPrepperEvent = JacksonEvent.builder()
-                    .withEventType(DATA_PREPPER_EVENT_TYPE)
-                    .withData(rowDataMap)
-                    .build();
-
-            final Event pipelineEvent = recordConverter.convert(
-                    dataPrepperEvent,
-                    tableMetadata.getDatabaseName(),
-                    tableMetadata.getTableName(),
-                    bulkAction,
-                    primaryKeys,
-                    eventTimestampMillis,
-                    eventTimestampMillis,
-                    event.getHeader().getEventType());
-            pipelineEvents.add(pipelineEvent);
+        for (Map.Entry<OpenSearchBulkActions, List<Serializable[]>> entry : bulkActionToRowDataMap.entrySet()) {
+            final OpenSearchBulkActions bulkAction = entry.getKey();
+            final List<Serializable[]> rows = entry.getValue();
```

Comment:
We should process all the events in order. Here looks like we process INDEX and DELETE operations separately. We might have a race condition where the same key might exist in INDEX and DELETE operation.

```
k1-> k2; k3->k1

Index keys: k2, k1
Delete keys: k1, k3
```
k1 is in both index and delete.

Reply:
Instead of `Map<OpenSearchBulkActions, List<Serializable[]>>`, we should have List<Object> that contains OpenSearchBulkAction as property to process the events in order.

Reply:
Good point. I took your suggestions and arrange the row change in order and delete before index. So in the above case, the generated events will be: delete k1, index k2, delete k3, index k1.

Reply:
Changes look good to me. Can you add unit test ?

Reply:
Thanks for adding the unit test!